test: use unique tmpdirs for each test

joaocgreis opened the pull request:
Some of the recent failures in CI have happened because tests leave processes running or somehow block the tmpdir, causing the next test that tries to use the same tmpdir to fail. We use one tmpdir per thread of the test runner.

This changes the test system to use a different directory for each test, removing interference between tests.

~With this, a test run will leave about 200 tmpdirs in `test/`, but these will be reused in future runs so they don't accumulate. Removing the tmpdir on process exit interferes with cleanup that some tests already do so it's not straightforward. For now, this doesn't create any false positives, only makes unrelated tests pass.~

cc @nodejs/testing 

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Trott commented:
Since our tests don't typically clean up the tmp dirs, this is going to leave a lot of tmp dirs lying around. Maybe it makes sense to add a tmpdir cleanup process in `test.py`?

Trott commented:
> Some of the recent failures in CI have happened because tests leave processes running or somehow block the tmpdir, causing the next test that tries to use the same tmpdir to fail. 

Is this not a bug that needs to be fixed? The first test should finish before the second test begins.

mscdex commented:
I agree, fixing the problematic test is better than keeping hundreds of temporary directories around.

joaocgreis commented:
Added code to clean the temp dir on process exit. Some tests already had cleanup on process exit that interferes with this, so I removed cleanup code from the tests. Also had to ensure every file descriptor is closed before the test completes.

The only test that actually requires a file to be left open at the end is `addons/load-long-path`, so I changed it to use a child process. The changes in other tests are fairly straightforward.

This makes the tests independent from each other concerning the temp dir. When a test leaves something running blocking the temp dir, **it will no longer cause some other test to fail** afterwards, but actually fail in the test that should fail. This will make it much easier to debug flaky tests.

Includes fixes for two Node issues that appeared while implementing this:
- Close the file descriptor for the promisified version of truncate: 119682e8b7fbc6de1c6c76d91b5a61428e22de05. The fix is similar to what is already there for `lchmod`, `writeFile` and `readFile`. @nodejs/fs 
- Close the file descriptor for the REPL history file: 7f06449daf40cc0db490d4a353cedc1b6a80b280. @nodejs/repl 

I can open these as separate PRs if needed. However, this PR would have to be blocked by those and the tests for the fixes is here.

It is tempting to clean the temp dirs on `test.py` instead of `tmpdir.js`, but I don't think that would be a good solution because: 1) `test.py` knows nothing about our temp dirs, it would be much better to keep this localized in `tmpdir.js`; 2) If there's any issue with the code in `test.py`, we break all the tests, not only the failing one; 3) I want to add some debug logging when removing the directory fails, ideally it should end up in the test report together with the error; 4) it wouldn't help us catch issues such as the ones above.

Trott commented on test/common/README.md:
```diff
@@ -906,6 +906,12 @@ The realpath of the testing temporary directory.
 
 Deletes and recreates the testing temporary directory.
 
+The first time refresh is run it adds a listener to process `'exit'` that
```
Optional nit:

```suggestion
The first time `refresh()` runs,  it adds a listener to process `'exit'` that
```

Trott commented on test/common/README.md:
```diff
@@ -906,6 +906,12 @@ The realpath of the testing temporary directory.
+The first time refresh is run it adds a listener to process `'exit'` that
+cleans the temporary directory. Thus, every file under `tmpdir.path` needs to
+be closed before the test completes. A good way to do this is to add a
+listener to process `'beforeExit'`. If a file needs to be left open until Node
```
Optional nit:

```suggestion
listener to process `'beforeExit'`. If a file needs to be left open until Node.js
```

or

```suggestion
listener to process `'beforeExit'`. If a file needs to be left open until `node`
```

Trott commented on test/common/README.md:
```diff
@@ -906,6 +906,12 @@ The realpath of the testing temporary directory.
+cleans the temporary directory. Thus, every file under `tmpdir.path` needs to
+be closed before the test completes. A good way to do this is to add a
+listener to process `'beforeExit'`. If a file needs to be left open until Node
+completes, use a child process and call `refresh` only in the parent.
```
Optional nit:

```suggestion
completes, use a child process and call `refresh()` only in the parent.
```

Trott commented:
I realize we don't run these tests on Windows in CI, but there are tests in `pummel` (such as `test-fs-largefile.js`) that have an exit handler doing `unlinkSync()`. Should those be adjusted too?

Trott commented:
I like this approach a lot. Thanks for doing this!

joaocgreis commented:
> tests in `pummel`... Should those be adjusted too?

Yes! Thanks for bringing this up!

Trott commented:
Looks like the main issue here is that the tmp directories on Raspberry Pi devices in CI are NFS mounted and that can mess with the cleanup logic.

One possible solution is to detect that it's an NFS file system and ignore cleanup errors in that situation. Another possibility is to ignore cleanup issues if they are on files of the format `/^\.nfs\w+$/` or (perhaps more robust and easier to maintain) `.startsWith('.nfs')`.

joaocgreis commented:
@nodejs/benchmarking added a commit (134424386c2881353fb1b4126f297d8d6ef32ff4) to switch 3 benchmarks to `common/tmpdir.js`, to be able to run the benchmark test suite. There are other benchmarks that already use this.

@bnoordhuis, @indutny, @nodejs/streams: the only issue left here is that `net.Server` does not seem to close the file descriptor when listening on a pipe. This happens in both [`parallel/test-child-process-server-close`](https://github.com/nodejs/node/blob/61f3a5c60ad78506e9e0caae061a04ccab878ca1/test/parallel/test-child-process-server-close.js) and [`parallel/test-tls-wrap-econnreset-pipe`](https://github.com/nodejs/node/blob/61f3a5c60ad78506e9e0caae061a04ccab878ca1/test/parallel/test-tls-wrap-econnreset-pipe.js). I confirmed by adding this to the tests:

```js
const { execSync } = require('child_process');
const lsof = `lsof -p ${process.pid} | grep -e ^COMMAND -e tmp`;
process.on('exit', () => {
  const openFiles = execSync(lsof, { encoding: 'utf8' });
  console.error(`Open files (${lsof}):\n${openFiles}`);
});
```

Is this a real issue in Node? Since the server is closed in both tests shouldn't the file descriptor be closed? Or is this by design and we should run those tests in child processes like [`parallel/test-pipe-unref`](https://github.com/JaneaSystems/node/commit/2367954d5b337c0f5c249e8f49e9f8ebe4485deb#diff-b983954bc5f11978614ed40d40fc4071) here?

@Trott I've only seen the `.nfs*` files when a file is actually open. This may be a real issue. For now, I've added a warning about this to the cleanup logic.

Trott commented:
It would perhaps not be the most elegant solution, but maybe if `.nfs*` files are detected and nothing else, wait and check again?

joaocgreis commented:
Updated. Changed those two tests to use a child process. With this solution we don't lose anything, the tests keep testing exactly what they were testing before.

This is ready for review (hopefully final this time).

@Trott the `.nfs*` files won't go away as long as the file descriptors are open. So, they won't go away and the test would timeout. The files are also left open when not using NFS, but don't belong to any directory so don't block removing the directory. I made some code to show this happening on Linux: https://github.com/JaneaSystems/node/commit/bbb42ef088566f229d2b10b805da90154855a722, but won't include it here because it makes the tests take 40 seconds longer to run (on my machine).

Trott commented:
@refack Any idea if this might make the [stuff you added a few months ago to the `rimraf` code in `common`](https://github.com/nodejs/node/blob/2b03e1d0a5ef2a52e4d7ce550b74a961c62e185a/test/common/tmpdir.js#L31) unnecessary?

bnoordhuis commented:
@joaocgreis Apropos https://github.com/nodejs/node/pull/28858#issuecomment-517036226, are you perhaps calling lsof too early? With the patch below no open file descriptors are reported (and indeed strace logs a close(2) call.)

```diff
diff --git a/test/parallel/test-child-process-server-close.js b/test/parallel/test-child-process-server-close.js
index d70926f2e8..588455b7da 100644
--- a/test/parallel/test-child-process-server-close.js
+++ b/test/parallel/test-child-process-server-close.js
@@ -17,7 +17,14 @@ const server = net.createServer((conn) => {
   const client = net.connect(common.PIPE, common.mustCall());
   client.on('data', () => {
     client.end(() => {
-      server.close();
+      server.close(() => {
+        const { execSync } = require('child_process');
+        const lsof = `lsof -p ${process.pid} | grep -e ^COMMAND -e tmp`;
+        process.on('exit', () => {
+            const openFiles = execSync(lsof, { encoding: 'utf8' });
+            console.error(`Open files (${lsof}):\n${openFiles}`);
+        });
+      });
     });
   });
 });
```

joaocgreis commented:
Rebased.

@Trott thanks for the review! I don't know exactly why the `rmdir` was added, I'm guessing as a broad-spectrum attempt to fix errors that were not very clear (hopefully this PR will help with that). For now it doesn't hurt (only makes the tests a bit slower). I've disabled it for the cleanup on exit, so that if errors appear we can fix them. If we see less issues from now on, perhaps we can remove `rmdir` some time in the future.

@bnoordhuis I can't get the close callback to run at all. Do you see the lsof headers, as proof that it ran? If I'm not missing anything here, I can open another issue about the callback not running, unless you think this PR should block on that.

refack commented on tools/test.py:
```diff
@@ -535,6 +539,7 @@ def RunCommand(self, command, env):
   def Run(self):
     try:
       result = self.RunCommand(self.GetCommand(), {
+        "TEST_SERIAL_ID": "%d" % self.serial_id,
```
IMO this should be something that is traceable to the tested file.
maybe something like:
```py
        "TEST_SERIAL_ID": "%d" % re.sub('\W', '-', self.path),
```
This should also eliminate the need for changes in `ProgressIndicator`

joaocgreis replied:
That's a good change, thanks! The number needs to stay for when using `--repeat`.

joaocgreis replied:
Had to take this back because of http://docs.libuv.org/en/v1.x/pipe.html#c.uv_pipe_bind.

Would be good to have the test name in the tmpdir, but I don't see an easy way around this. With this PR, the tmpdir either gets removed correctly or the test that is using it will fail and print information. In any case, it's something that we didn't have before.

@refack let me know if you feel strongly about this and have some suggestion.

refack commented:
> Some of the recent failures in CI have happened because tests leave processes running or somehow block the tmpdir,

My intuition has been that this is caused by a real underling bug related to how node detects and deals with exit status of child processes on Windows...

> Any idea if this might make the [stuff you added a few months ago to the `rimraf` code in `common`](https://github.com/nodejs/node/blob/2b03e1d0a5ef2a52e4d7ce550b74a961c62e185a/test/common/tmpdir.js#L31) unnecessary?

I was hoping that a call to `rmdir` would better handle Windows specific issues like files locked by the Indexing Service or Windows Defender.

joaocgreis commented:
Updated with suggestion.

Added the semver-major label. Perhaps not strictly needed, but it's better to be careful. cc @nodejs/tsc

> My intuition has been that this is caused by a real underling bug related to how node detects and deals with exit status of child processes on Windows...

I suspect `test.py` more than `node`. In either case, this PR will make the error happen on the test that actually has the issue instead of some random test later.

> I was hoping that a call to `rmdir` would better handle Windows specific issues like files locked by the Indexing Service or Windows Defender.

Makes sense. Those two are the usual suspects and I suspect issues are attributed to them more often than they should, so I'd like a chance to investigate actual failures after this PR lands.

joaocgreis commented:
CitGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1926/ (waiting in queue)

joaocgreis commented:
When running in CI with `NODE_TEST_DIR` defined, this hits the limitation described in http://docs.libuv.org/en/v1.x/pipe.html#c.uv_pipe_bind. I've removed the last fixup commit which added the test name to the temp dir as suggested by @refack.

CI: https://ci.nodejs.org/job/node-test-pull-request/24865/ ✔️

Trott commented:
A commit just landed on master that will make CITGM results easier to compare because it greatly reduces failures. So a post-force-push CITGM run would probably be a good idea, if you think this is in a final or near-final state.

joaocgreis commented:
CitGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1933/

joaocgreis commented:
@bnoordhuis opened an issue about the `net.Server` tests: https://github.com/nodejs/node/issues/29069

I don't think this PR should be blocked by those tests, the change in the last commit makes the tests pass and work exactly as before.

Trott commented:
@nodejs/tsc This needs at least one more TSC approval. (@cjihrig maybe?)

Trott commented:
Landed in df936c5...8ef68e6

cclauss commented:
Impressive work here!  Congratulations.
